For E0277 on `for` loops, point at the "head" expression

estebank opened the pull request:
When E0277's span points at a `for` loop, the actual issue is in the
element being iterated. Instead of pointing at the entire loop, point
only at the first line (when possible) so that the span ends in the
element for which E0277 was triggered.

estebank committed:
For E0277 on `for` loops, point at first line

When E0277's span points at a `for` loop, the actual issue is in the
element being iterated. Instead of pointing at the entire loop, point
only at the first line (when possible) so that the span ends in the
element for which E0277 was triggered.

rust-highfive commented:
r? @eddyb

(rust_highfive has picked a reviewer for you, use r? to override)

eddyb commented:
r? @nikomatsakis

zackmdavis commented on src/librustc/traits/error_reporting.rs:
```diff
@@ -551,6 +551,24 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
                             = self.on_unimplemented_note(trait_ref, obligation);
                         let have_alt_message = message.is_some() || label.is_some();
+                        let span = match self.tcx.sess.codemap().span_to_snippet(span) {
+                            Ok(ref _s) if _s.starts_with("for ") => {
```
style nit: `Ok(ref s) if s.starts_with("for ")` ([pretty sure](https://play.rust-lang.org/?gist=3fd8dec472b33b02f0b34e92e1bdec15&version=stable) the unused-variables lint is in fact smart enough to count the `if` guard as a use)

estebank replied:
For some reason I was convinced that there was a lint that complained at patterns that where only used in match guards... Must have been a different case.

zackmdavis replied:
You might be thinking of how the dead-code lint complains about enum variants that are never constructed, even if they're matched on?

nikomatsakis commented on src/librustc/traits/error_reporting.rs:
```diff
@@ -551,6 +551,24 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
                         let have_alt_message = message.is_some() || label.is_some();
+                        let span = match self.tcx.sess.codemap().span_to_snippet(span) {
+                            Ok(ref s) if s.starts_with("for ") => {
+                                // On for loops, this error is caused by the element being iterated
```
Hmm, I like this span a lot, but I wonder if this is the right way to do this. Maybe we should adjust the span in HIR lowering instead? Roughly here:

https://github.com/rust-lang/rust/blob/a0dcecff90c45ad5d4eb60859e22bb3f1b03842a/src/librustc/hir/lowering.rs#L3167-L3173

estebank replied:
Reasonable, it's definitely an overuse of `span_until_char`, it was just slightly faster to figure how to do this than the proper way. Should it point at only the iterator expression or at the entire `for _ in iterator_expr`?

nikomatsakis replied:
@estebank good question, I don't know the answer. Just pointing at the span of the iterator expression is obviously *easier*... and maybe better? I guess that showing the `for` may help convey that it's "because of the `for`" that `IntoIterator` is needed... but that seems sort of obvious...

If we do it in lowering, we can also introduce a "synthetic" span, right, that would let us give a very tailored message if we wanted?

estebank replied:
> If we do it in lowering, we can also introduce a "synthetic" span, right, that would let us give a very tailored message if we wanted?

I think so. If we want to point at the `for`, we could just do that, which would look better ni the case where `for` and `expr` are not in the same line:

```
X |    for _ in expr {
  |    ---      ^^^^ `&str` is not an iterator
  |    |
  |    needed because of this `for` loop
```

nikomatsakis replied:
oooh, that's nice

estebank replied:
I looked at this and I have to rework the code to not lose the for-loop's context by the time the trait is being resolved. It might take me a while.

estebank replied:
Modified the expansion code to use the appropriate span, but that makes it harder to add the appropriate span under `for`.

nikomatsakis reviewed (changes requested):
I like the span but feel like we could do it in a more elegant way...

Also, what if people do something like

```
for x in
    something
{
    ...
}
```

I do that from time to time...

estebank committed:
Modify spans of expanded expression

Modify the spans used for `for`-loop expression expansion, instead of
creating a new span during error creation.

nikomatsakis commented on src/librustc/hir/lowering.rs:
```diff
@@ -2969,7 +2969,7 @@ impl<'a> LoweringContext<'a> {
                         P(self.lower_expr(else_expr))
                     } else {
-                        self.expr_tuple(e.span, hir_vec![])
+                        self.expr_tuple(body.span, hir_vec![])
```
Why the change to body.span here? It's not entirely intuitive to me that the (synthetic) "else clause" should get the span of the code that executes when `else` does not apply. =) At minimum, a comment seems good.

estebank replied:
Reverted.

nikomatsakis commented on src/test/ui/const-fn-error.stderr:
```diff
@@ -13,22 +13,20 @@ error[E0016]: blocks in constant functions are limited to items and tail express
+17 |     for i in 0..x { //~ ERROR calls in constant functions
+   |              ^^^^
+   |              |
+   |              in this macro invocation
```
macro invocation? what macro? =)

Is this referring to the for loop desugaring?

estebank replied:
I believe it is the desugaring of the range. Is it ok if I tackle this on a follow up PR? The span is probably pointing at the `x`.

estebank replied:
Fixed.

nikomatsakis commented on src/test/ui/suggestions/for-c-in-str.rs:
```diff
@@ -0,0 +1,21 @@
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// E0277 should point exclusively at line 14, not the entire for loop span
```
nice

estebank committed:
Don't add "in this macro invocation" label to desugared spans

nikomatsakis commented on src/test/ui/suggestions/try-operator-on-main.stderr:
```diff
@@ -2,10 +2,7 @@ error[E0277]: the `?` operator can only be used in a function that returns `Resu
-   |     |
-   |     cannot use the `?` operator in a function that returns `()`
-   |     in this macro invocation
+   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot use the `?` operator in a function that returns `()`
```
nice drive-by fix here 🚗

estebank replied:
😎

nikomatsakis commented on src/test/ui/suggestions/for-c-in-str.stderr:
```diff
@@ -0,0 +1,11 @@
+   |              ^^^^^^ `&str` is not an iterator; maybe try calling `.iter()` or a similar method
+   |
+   = help: the trait `std::iter::Iterator` is not implemented for `&str`
+   = note: required by `std::iter::IntoIterator::into_iter`
```
it'd be cool if we could highlight the `for`, but this is indeed pretty good still =)

estebank replied:
I agree. That will require some extra changes to `report_selection_error` in order to figure out that this is part of a `for`-loop desugaring in order to find the leading `for`. For now I think this is enough of an improvement.

nikomatsakis reviewed (approved):
this turned out really good

nikomatsakis commented:
@bors r+

bors commented:
:pushpin: Commit 106e5c5 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 106e5c554d6b6b97aecac254a2694247e84e718e -->

bors commented:
:hourglass: Testing commit 106e5c554d6b6b97aecac254a2694247e84e718e with merge 6272b60dcaae2e69c991fc64cfb2e403c8a616a3...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.6160), [status-travis](https://travis-ci.org/rust-lang/rust/builds/334040846?utm_source=github_status&utm_medium=notification)
Approved by: nikomatsakis
Pushing 6272b60dcaae2e69c991fc64cfb2e403c8a616a3 to master...
